src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/FeatureSwitchManager.cs

```diff
@@ -139,9 +146,17 @@ public MethodIL GetMethodILWithInlinedSubstitutions(MethodIL method)
             // Last step is a sweep - we replace the tail of all unreachable blocks with "br $-2"
             // and nop out the rest. If the basic block is smaller than 2 bytes, we don't touch it.
             // We also eliminate any EH records that correspond to the stubbed out basic block.
+            //
+            // We also attempt to rewrite calls to SR.SomeResourceString accessors with string
+            // literals looked up from the managed resources.
 
             Debug.Assert(method.GetMethodILDefinition() == method);
 
+            // Do not attempt to inline resource strings if we only want to use resource keys.
+            // The optimizations are not compatible.
+            bool shouldInlineResourceStrings =
+                !_hashtable._switchValues.TryGetValue("System.Resources.UseSystemResourceKeys", out bool useResourceKeys) || !useResourceKeys;
```

Comment:
Any estimate on additional saving we would get if this feature also worked with `UseSystemResourceKeys` feature switch (some of the message are very long)?

Reply:
> Any estimate on additional saving we would get if this feature also worked with `UseSystemResourceKeys` feature switch (some of the message are very long)?

This will always be inlined - for the below app:

```csharp
Console.WriteLine(new NullReferenceException().Message);
throw null;
```

If I compile this with .NET 7 as `dotnet publish -c Release /p:PublishAot=true /p:UseSystemResourceKeys=true` and step into the ctor, I see this (the first `lea` is loading the string literal of the resource key - inlined from the accessor):

```
00007FF787207950 56                   push        rsi  
00007FF787207951 48 83 EC 20          sub         rsp,20h  
00007FF787207955 48 8B F1             mov         rsi,rcx  
00007FF787207958 48 8D 0D 11 7B 17 00 lea         rcx,[__Str_Arg_NullReferenceException (07FF78737F470h)]  
00007FF78720795F E8 0C C2 00 00       call        S_P_CoreLib_System_SR__GetResourceString (07FF787213B70h)  
00007FF787207964 C7 46 48 00 15 13 80 mov         dword ptr [rsi+48h],80131500h  
00007FF78720796B 48 8D 4E 08          lea         rcx,[rsi+8]  
00007FF78720796F 48 8B D0             mov         rdx,rax  
00007FF787207972 E8 89 BE F4 FF       call        RhpAssignRefAVLocation (07FF787153800h)  
00007FF787207977 C7 46 48 01 15 13 80 mov         dword ptr [rsi+48h],80131501h  
00007FF78720797E C7 46 48 03 40 00 80 mov         dword ptr [rsi+48h],80004003h  
00007FF787207985 48 83 C4 20          add         rsp,20h  
00007FF787207989 5E                   pop         rsi  
00007FF78720798A C3                   ret  
```

`SR__GetResourceString` didn't get inlined probably because the IL is messed up with too many nops and looks large even thought it isn't. It looks like this and there's only one of it, so not a huge deal size-wise, but we could leave fewer nops when substitutions happen to not to throw off inlining heuristics:

```
00007FF787213B70 48 8B C1             mov         rax,rcx  
00007FF787213B73 C3                   ret  
```